src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/common/DirectoryControl.cs

```diff
@@ -132,102 +154,165 @@ public virtual byte[] GetValue()
+                            asnSpan = asnSpan.Slice(bytesConsumed);
+                        }
+                        ThrowUnless(asnSpan.IsEmpty);
 
-                    // decode might fail as AD for example never returns attribute name, we don't want to unnecessarily throw and catch exception
-                    if (decodeSucceeded)
-                    {
-                        Debug.Assert(o != null && o.Length == 2);
-                        result = (int)o[0];
-                        attribute = (string)o[1];
+                        PageResultResponseControl pageControl = new PageResultResponseControl(size, cookie, controls[i].IsCritical, value);
+                        controls[i] = pageControl;
                     }
-                    else
+                    else if (controls[i].Type == "1.2.840.113556.1.4.1504")
                     {
-                        // decoding might fail as attribute is optional
-                        o = BerConverter.Decode("{e}", value);
-                        Debug.Assert(o != null && o.Length == 1);
+                        // The control is an AsqControl, as described in MS-ADTS section 3.1.1.3.4.1.18
+                        AsnDecoder.ReadSequence(asnSpan, AsnEncodingRules.BER, out int sequenceContentOffset, out int sequenceContentLength, out _);
+                        ThrowUnless(sequenceContentLength > 0);
 
-                        result = (int)o[0];
-                    }
+                        ResultCode result = AsnDecoder.ReadEnumeratedValue<ResultCode>(asnSpan.Slice(sequenceContentOffset, sequenceContentLength), AsnEncodingRules.BER, out _);
 
-                    SortResponseControl sort = new SortResponseControl((ResultCode)result, attribute, controls[i].IsCritical, controls[i].GetValue());
-                    controls[i] = sort;
-                }
-                else if (controls[i].Type == "2.16.840.1.113730.3.4.10")
-                {
-                    // The control is a VlvResponseControl.
-                    int position;
-                    int count;
-                    int result;
-                    byte[] context = null;
-                    object[] o = BerConverter.TryDecode("{iieO}", value, out bool decodeSucceeded);
-
-                    if (decodeSucceeded)
+                        AsqResponseControl asq = new AsqResponseControl(result, controls[i].IsCritical, value);
+                        controls[i] = asq;
+                    }
+                    else if (controls[i].Type == "1.2.840.113556.1.4.841")
                     {
-                        Debug.Assert(o != null && o.Length == 4);
-                        position = (int)o[0];
-                        count = (int)o[1];
-                        result = (int)o[2];
-                        context = (byte[])o[3];
+                        // The control is a DirSyncControl, as described in MS-ADTS section 3.1.1.3.4.1.3
```

Comment:
`DirSyncControl` does not appear in that document (in revision 60).  Do you mean `DirSyncRequestValue`?